kani-compiler/src/kani_middle/transform/loop_contracts.rs

```diff
@@ -0,0 +1,313 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains code related to the MIR-to-MIR pass to enable loop contracts.
+//!
+
+use crate::kani_middle::KaniAttributes;
+use crate::kani_middle::codegen_units::CodegenUnit;
+use crate::kani_middle::find_fn_def;
+use crate::kani_middle::transform::TransformationType;
+use crate::kani_middle::transform::body::{InsertPosition, MutableBody, SourceInstruction};
+use crate::kani_queries::QueryDb;
+use crate::stable_mir::CrateDef;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Symbol;
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::{
+    AggregateKind, BasicBlockIdx, Body, ConstOperand, Operand, Place, Rvalue, StatementKind,
+    Terminator, TerminatorKind, VarDebugInfoContents,
+};
+use stable_mir::ty::{FnDef, MirConst, RigidTy};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::fmt::Debug;
+
+use super::TransformPass;
+
+/// This pass will perform the following operations:
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
+///    to invoke the closure.
+///
+/// 2. Transform loops with contracts from
+///    ```ignore
+///    bb_idx: {
+///         loop_head_stmts
+///         _v = kani_register_loop_contract(move args) -> [return: terminator_target];
+///    }
+///
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_idx;
+///    }
+///    ```
+///    to blocks
+///    ```ignore
+///    bb_idx: {
+///         _v = true
+///         goto -> terminator_target
+///    }
+///
+///    ...
+///    loop_body_blocks
+///    ...
+///
+///    loop_latch_block: {
+///         loop_latch_stmts
+///         goto -> bb_new_loop_latch;
+///    }
+///
+///    bb_new_loop_latch: {
+///         loop_head_body
+///         _v = kani_register_loop_contract(move args) -> [return: terminator_target];
+///    }
+///    ```
+#[derive(Debug, Default)]
+pub struct LoopContractPass {
+    /// Cache KaniRunContract function used to implement contracts.
+    run_contract_fn: Option<FnDef>,
+    /// The map from original loop head to the new loop latch.
+    /// We use this map to redirect all original loop latches to a new single loop latch.
+    new_loop_latches: HashMap<usize, usize>,
```

Comment:
You can't cache this here. This object will get re-used for all body transformations. If you need to cache something that is relevant to a specific transformation, you should create a new object that only lives during the transformation of a function.